### PR TITLE
IGNITE-19234 Enable and fix group reentry logic for volatile storages

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-19234

The reentry logic (see IGNITE-16668) was erroneously disabled. This commit enables it back and contains fixes required to make it work (due to changes made to the common rebalancing code since it was disabled).

### CODE REVIEW

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/message/DataPresence.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.table.distributed.message;
+
+/**
+ * Whether a node has data or not (or it's not known).
+ */
+public enum DataPresence {
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/message/DataPresence.java:23]: Can we serialize it into an int, not a string?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/message/DataPresence.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.table.distributed.message;
+
+/**
+ * Whether a node has data or not (or it's not known).
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/message/DataPresence.java:21]: Can you please clarify a bit, what does "not known" mean?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/PartitionReplicatorNodeRecovery.java [hunk 3]
```diff
@@ -285,12 +314,30 @@ private CompletableFuture<Long> queryDataNodesCount(Collection<Peer> peers, HasD
...
+                            .collect(toList());
+
+                    long nodesSurelyHavingData = hasDataFlags.stream().filter(presence -> presence == DataPresence.HAS_DATA).count();
+                    long nodesSurelyEmpty = hasDataFlags.stream().filter(presence -> presence == DataPresence.EMPTY).count();
+                    return new DataNodesCounts(nodesSurelyHavingData, nodesSurelyEmpty);
+                });
+    }
+
+    private static class DataNodesCounts {
+        private final long nodesSurelyHavingData;
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/PartitionReplicatorNodeRecovery.java:335]: These fields' names are a bit of a mouthful. Can we rename them to something like "dataNodesCnt" and "emptyNodesCnt"?

Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/PartitionReplicatorNodeRecovery.java:335]: Is it a problem? As a result, the code that uses them is self-explanatory. On the other hand, `dataNodesCnt` does not tell you that we actually checked that there is the given count of replicas *available*.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/message/DataPresence.java (hunk 1 above)
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/message/DataPresence.java:24]: I don't mean to use ordinals, I mean to use an explicit int value. Don't we have a similar approach in other enums? Anyway, that's up to you

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/PartitionReplicatorNodeRecovery.java (hunk 3 above)
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/PartitionReplicatorNodeRecovery.java:335]: For me, actually, the current naming was very confusing for some reason. I had to re-read them multiple times. If you don't like my suggestion, maybe you could provide something simpler?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/message/DataPresence.java (hunk 2 above)
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/message/DataPresence.java]: Added clarifications

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/message/DataPresence.java (hunk 1 above)
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/message/DataPresence.java:24]: Maybe I just missed this, but I did not see this approach in use for network messages. I still prefer strings for the reason I described earlier when the performance penalty is small (like in this case).

**Review by sashapolo — APPROVED** [review]

**Pull request merged**